File: PriceFeed-Kaiko/src/oracle.js
const https   = require('https');
const ethers  = require('ethers');
const fs      = require('fs');

const root                = 'iexec_out';
const determinismFilePath = `${root}/determinism.iexec`;
const callbackFilePath    = `${root}/callback.iexec`;
const errorFilePath       = `${root}/error.iexec`;

/*****************************************************************************
 *                                  CONFIG                                   *
 *****************************************************************************/
// const EXCHANGE = 'cbse'
const EXCHANGE = 'bnce'
const APIKEY   = '';

const WAIT_MIN = parseInt(process.env.WAIT_MIN) || 0; // in ms
const WAIT_MAX = parseInt(process.env.WAIT_MAX) || 0; // in ms

/*****************************************************************************
 *                                   TOOLS                                   *
 *****************************************************************************/
process.sleep = (ms) => {
	return new Promise(resolve => setTimeout(resolve, ms));
}

Math.average = (input) => {
	this.output = 0;
	for (this.i = 0; this.i < input.length; ++this.i)
	{
		this.output+=Number(input[this.i]);
	}
	return this.output/input.length;
}

/*****************************************************************************
 *                                 ARGUMENTS                                 *
 *****************************************************************************/

var [ asset_id_base, asset_id_quote, power, time ] = process.argv.slice(2);
if (/^\d*$/.test(time)) { time = new Date(parseInt(time)*1000).toISOString(); }

// const asset_id_base  = "BTC"
// const asset_id_quote = "USD"
// const power          = 9
// const time           = new Date().toISOString();

/*****************************************************************************
 *                                HTTP QUERY                                 *
 *****************************************************************************/

const fragment = Object.entries({
	// 'interval': '1m',
	'start_time': time,
}).filter(([k,v]) => v).map(([k,v]) => `${k}=${v}`).join('&');

const query = {
	method: 'GET',
	port: 443,
	host: 'eu.market-api.kaiko.io',
	// path: `/v1/data/trades.v1/exchanges/${EXCHANGE}/spot/${asset_id_base}-${asset_id_quote}/aggregations/vwap?${fragment}`,
	path: `/v1/data/trades.v1/exchanges/${EXCHANGE}/spot/${asset_id_base}-${asset_id_quote}/trades?${fragment}`,
	headers: {
		'Accept': 'application/json',
		'X-Api-Key': process.env.APIKEY || APIKEY,
	},
};

/*****************************************************************************
 *                                  EXECUTE                                  *
 *****************************************************************************/
new Promise(async (resolve, reject) => {

	const delay = (WAIT_MAX-WAIT_MIN) * Math.random() + WAIT_MIN;
	console.log(`- Waiting for ${delay} ms.`);
	await process.sleep(delay);

	console.log('- Calling API');
	var data = "";
	var request = https.request(query, res => {
		res.on('data', (chunk) => {
			data += chunk;
		});
		res.on('end', () => {
			if (data)
			{
				console.log('- Got data');
				resolve(data);
			}
			else
			{
				reject(`[HTTP ERROR]\nstatusCode: ${res.statusCode}`);
			}
		});
	});
	request.on('error', reject);
	request.end();
})
.then(data => {
	var results = JSON.parse(data.toString());

	if (results.result !== 'success')
	{
		throw new Error(results.message);
	}

	var timestamp = new Date(results.query.start_time).getTime();
	var details   = [ results.query.instrument, power].join('-').toUpperCase();
	var value     = Math.round(Math.average(results.data.map(t => t.price)) * 10**power);

	if (isNaN(timestamp) || isNaN(value) || results.asset_id_base  == "" || results.asset_id_quote == "")
	{
		throw new Error("Error: invalid results " + JSON.stringify({query, results}));
	}

	var iexeccallback = ethers.utils.defaultAbiCoder.encode(['uint256', 'string', 'uint256'], [timestamp, details, value]);
	var iexecconsensus = ethers.utils.keccak256(iexeccallback);
	fs.writeFile(callbackFilePath,    iexeccallback , (err) => {});
	fs.writeFile(determinismFilePath, iexecconsensus, (err) => {});

	console.log(`- Success: ${timestamp} ${details} ${value}`);
})
.catch(error => {
	fs.writeFile(
		errorFilePath,
		error.toString(),
		(err) => {}
	);
	fs.writeFile(
		determinismFilePath,
		ethers.utils.solidityKeccak256(['string'],[error.toString()]),
		(err) => {}
	);
	console.log(error.toString());
});

File: PriceFeed-Kaiko/app/oracle.py
#!/usr/bin/python3

import datetime
import eth_abi
import json
import jsonpath_rw
import os
import re
import sha3
import sys
import urllib.request

root         = '/'
inputDir     = '{}iexec_in/'.format(root)
outputDir    = '{}scone/'.format(root)
callbackFile = 'callback.iexec'

class Lib:
	def parseValue(rawValue, ethType, power):
		if re.search('^u?int[0-9]*$', ethType):
			return round(float(rawValue) * 10 ** int(power))
		else:
			return rawValue

	def formatArgs(args):
		return '&'.join('{}={}'.format(k,v) for k,v in args.items())

	def getAPIKey():
		# Dataset file is a zip that is extracted before the entrypoint. read the key.txt file extracted from it
		# file = os.environ.get('IEXEC_DATASET_FILENAME')
		file = 'key.txt'
		path = '{root}{file}'.format(root=inputDir, file=file)
		try:
			with open(path, 'r') as file:
				apiKey = file.read().strip()
				if not re.search('^[0-9a-zA-Z]{1,128}$', apiKey):
					raise Exception('Invalid API key')
				return apiKey
		except FileNotFoundError:
			raise Exception('Missing API key dataset')

	def fetchMarketData(region, endpoint, params):
		print('Request https://{region}.market-api.kaiko.io/v1/data/trades.v1/{endpoint}?{params}'.format(
			region   = region,
			endpoint = endpoint,
			params   = params,
		))
		return json.loads(
			urllib.request.urlopen(
				urllib.request.Request(
					'https://{region}.market-api.kaiko.io/v1/data/trades.v1/{endpoint}?{params}'.format(
						region   = region,
						endpoint = endpoint,
						params   = params,
					),
					headers = {
						'X-Api-Key': Lib.getAPIKey(),
						'User-Agent': 'Kaiko iExec Adapter',
					}
				)
			).read()
		)

class PriceFeed:
	def fetchRate(baseAsset, quoteAsset):
		return Lib.fetchMarketData(
			'us',
			'spot_direct_exchange_rate/{baseAsset}/{quoteAsset}/recent'.format(baseAsset=baseAsset, quoteAsset=quoteAsset),
			Lib.formatArgs({
				'interval': '1m',
				'limit':    720,
			})
		)

	def run(baseAsset, quoteAsset, power):
		response = PriceFeed.fetchRate(
			baseAsset  = baseAsset,
			quoteAsset = quoteAsset,
		)
		try:
			data      = response.get('data')[0]
			timestamp = data.get('timestamp')
			details   = '-'.join([ baseAsset, quoteAsset, power ])
			rawValue  = data.get('price')
			value     = Lib.parseValue(rawValue, 'uint256', power)
			return (timestamp, details, value)
		except Exception as e:
			raise Exception('API response parsing failure: {}'.format(e))


# Example usage:
# btc usd 9
if __name__ == '__main__':
	print('PriceFeed started')
	try:
		# EXECUTE CALL
		result = PriceFeed.run(
			baseAsset  = sys.argv[1],
			quoteAsset = sys.argv[2],
			power      = sys.argv[3],
		)
		print('- Success: {} {} {}'.format(*result))

		# GENERATE CALLBACK
		callback = eth_abi.encode_abi([ 'uint256', 'string', 'uint256' ], [ *result ]).hex()
		with open('{path}{file}'.format(path=outputDir, file=callbackFile), 'w') as file:
			file.write('0x{}'.format(callback))

	except IndexError as e:
		print('Error: missing arguments')

	except Exception as e:
		print('Execution Failure: {}'.format(e))

	print('PriceFeed completed')
